### ISSUE TITLE
history-substring-search defines function that collides with beep alias

### ISSUE THREAD

**Issue by jbulchergist** [robbyrussell/oh-my-zsh#3913]
I would love to activate ZLE and get rid of this:

```
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
```

... every time I hit enter. I changed systems from Ubuntu to openSUSE, and now zsh complains about my custom theme. How can I activate ZLE? There doesn't seem to be any documentation on google or any of the zsh, oh-my-zsh projects. I'm not even sure if this is a zsh problem or an .oh-my-zsh problem.

I'd also like to know if there is a way to check to see if the option is set; the documentation for zshzle indicates the option is set by default in interactive shells, so I don't know why it wouldn't be set in mine.

I have no idea where to start troubleshooting this. Help!

-Thanks

**Comment by mcornella** [comment]
You're probably using a zle widget inside a function that gets called everytime you update your prompt. Please post your full custom theme to see what we're dealing with.

**Comment by jbulchergist** [comment]
Thanks for helping with this! Here you go:

```sh
# Name:   trapd00r zsh theme
# Author: Magnus Woldrich <m@japh.se>
#
# This theme needs a terminal supporting 256 colors as well as unicode. It also
# needs the script that splits up the current path and makes it fancy as located
# here: https://github.com/trapd00r/utils/blob/master/zsh_path
#
# By default it spans over two lines like so:
#
# scp1@shiva:pts/9-> /home » scp1 (0)
# >
#
# that's  user@host:pts/-> splitted path (return status)
#
# If the current directory is a git repository, we span 3 lines;
#
# git❨ master ❩ DIRTY
# scp1@shiva:pts/4-> /home » scp1 » dev » utils (0)
# >

autoload -Uz vcs_info
autoload -U add-zsh-hook

[ -n "$TMUX" ] && export TERM=screen-256color

#prompt_lvl="𝄞%"
prompt_lvl="$"
if $PUTTY ; then prompt_lvl="$"; fi

# We dont want to use the extended colorset in the TTY / VC.
if [ "$TERM" = "linux" ]; then
  source $ZSH/themes/afowler.zsh-theme
  return
fi


local c0='%{\e[m%}'

# colors
local c1="%{$(printf "\e[38;5;231m")%}" # white
local c3="%{$(printf "\e[38;5;250m")%}" # gray-
local c5="%{$(printf "\e[38;5;245m")%}" # gray
local c7="%{$(printf "\e[38;5;240m")%}" # gray+
local c9="%{$(printf "\e[38;5;322m")%}" # black

local c11="%{$(printf "\e[38;5;208m")%}" # orange
local c12="%{$(printf "\e[38;5;142m")%}" # yellow-green
local c13="%{$(printf "\e[38;5;160m")%}" # red
local c14="%{$(printf "\e[38;5;112m")%}" # green
local c15="%{$(printf "\e[38;5;230m")%}" # light yellow


# background colors
# local bc1="%{$(printf "\e[48;5;321m")%}" # white
# local bc3="%{$(printf "\e[48;5;250m")%}" # gray-
# local bc5="%{$(printf "\e[48;5;245m")%}" # gray
# local bc7="%{$(printf "\e[48;5;240m")%}" # gray+
# local bc9="%{$(printf "\e[48;5;322m")%}" # black

local bc11="%{$(printf "\e[48;5;208m")%}" # orange
# local bc12="%{$(printf "\e[48;5;142m")%}" # yellow-green
# local bc13="%{$(printf "\e[48;5;160m")%}" # red
# local bc14="%{$(printf "\e[48;5;112m")%}" # green
local bc15="%{$(printf "\e[48;5;230m")%}" # light yellow

zstyle ':vcs_info:*' actionformats \
    '%{$c8%}(%f%s)%{$c7%}-%F{5}[%F{2}%b%F{3}|%F{1}%a%F{5}]%f '

zstyle ':vcs_info:*' formats \
    "%{$c8%}%s%%{$c7%}❨ %{$c9%}%{$c11%}%b%{$c7%} ❩%{$reset_color%}%f "


zstyle ':vcs_info:(sv[nk]|bzr):*' branchformat '%b%F{1}:%F{3}%r'
zstyle ':vcs_info:*' enable git

add-zsh-hook precmd prompt_jnrowe_precmd

# Get the status of the working tree
custom_git_prompt_status() {
  STATUS=$(git_prompt_status)
  if [[ -n "$STATUS" ]]; then
    STATUS="$ZSH_THEME_GIT_PROMPT_STATUS_SEPARATOR$STATUS"
  fi
  echo $STATUS
}

prompt_jnrowe_precmd () {
  vcs_info
  local prompt_git_dir
  dir_status="$(pwd | sed "s@${HOME}@~@g")"
  dir_status="$(echo ${dir_status} | sed "s@/@%{$c11%}/%{$c14%}@g")"
  dir_status="%{$c14%}${dir_status}%{$reset_color%}"

  prompt_git_dir="$(git rev-parse --show-prefix 2> /dev/null)"
  prompt_git_dir=${prompt_git_dir%?}


  prompt_git_home=$(git rev-parse --show-toplevel 2> /dev/null)
  if [ "" = "${prompt_git_home}" ]; then
    prompt_git_home="$(pwd | sed "s@${HOME}@~@g")"
  else
    if [ "" != "$prompt_git_dir" ]; then
      prompt_git_dir="/${prompt_git_dir}"
    fi
    prompt_git_home=${prompt_git_home/${~HOME}/"~"}
    local top="%B${prompt_git_home:t}%b"
    prompt_git_home="${prompt_git_home:h}/${top}${prompt_git_dir}"
    prompt_git_dir="${top}${prompt_git_dir}"
  fi

  # color the directory dividers
  prompt_git_home="$(echo ${prompt_git_home} | sed "s@/@$c11/$c14@g")"
  prompt_git_home="%{$c14%}${prompt_git_home}%{$reset_color%}"
  prompt_git_dir="$(echo ${prompt_git_dir} | sed "s@/@$c11/$c14@g")"
  prompt_git_dir="%{$c14%}${prompt_git_dir}%{$reset_color%}"

  user_and_pc="%{$c5%}%n%B%{$c11%}@%b%{$c3%}%m%{$reset_color%}%B%{$c11%}:%b"

  if [[ $EUID -ne 0 ]]; then
    my_prompt="%B%{$c11%}${prompt_lvl}%b%{$reset_color%} "
  else
    my_prompt="%B%{$c11%}#%b%{$reset_color%} "
  fi

  ZSH_THEME_GIT_PROMPT_SUFFIX="%B%b$(custom_git_prompt_status) %{$c0%}%{$c15%}%B%b${prompt_git_dir}%{$c14%}%B%{$c11%}± %b%{$reset_color%}"
}
ZSH_THEME_GIT_PROMPT_PREFIX="
%{$fg[black]%}%{${bc15}%}%B "

ZSH_THEME_GIT_PROMPT_DIRTY="%B%{$c13%}✱ %b" # ❗❕☔☂Ⓓ
ZSH_THEME_GIT_PROMPT_CLEAN="%B%{$c14%}✔ %b" # ✔✅☀Ⓞ

ZSH_THEME_GIT_PROMPT_UNTRACKED="%{$fg[cyan]%} ?" # ✭ⓣ⬒⬓
ZSH_THEME_GIT_PROMPT_ADDED="%{$fg[cyan]%} ✚" # ⓐ ⑃
ZSH_THEME_GIT_PROMPT_MODIFIED="%{$c12%} ⚡"  # ⓜ ⑁
ZSH_THEME_GIT_PROMPT_DELETED="%{$c13%} ✖" # ⓧ ⑂
ZSH_THEME_GIT_PROMPT_RENAMED="%{$fg[blue]%} ➜" # ⓡ ⑄
ZSH_THEME_GIT_PROMPT_UNMERGED="%{$fg[magenta]%} ♒" # ⓤ ⑊

ZSH_THEME_GIT_PROMPT_STATUS_SEPARATOR=""

# More symbols to choose from:
# ☀ ✹ ☄ ♆ ♀ ♁ ♐ ♇ ♈ ♉ ♚ ♛ ♜ ♝ ♞ ♟ ♠ ♣ ⚢ ⚲ ⚳ ⚴ ⚥ ⚤ ⚦ ⚒ ⚑ ⚐ ♺ ♻ ♼ ☰ ☱ ☲ ☳ ☴ ☵ ☶ ☷
# ✡ ✔ ✖ ✚ ✱ ✤ ✦ ❤ ➜ ➟ ➼ ✂ ✎ ✐ ⨀ ⨁ ⨂ ⨍ ⨎ ⨏ ⨷ ⩚ ⩛ ⩡ ⩱ ⩲ ⩵  ⩶ ⨠
# ⬅ ⬆ ⬇ ⬈ ⬉ ⬊ ⬋ ⬒ ⬓ ⬔ ⬕ ⬖ ⬗ ⬘ ⬙ ⬟  ⬤ 〒 ǀ ǁ ǂ ĭ Ť Ŧ
# ‼ • ‣

# ${user_and_pc}${dir_status}%{$reset_color%}${my_prompt}$(git_prompt_info)%{$reset_color%}%{$reset_color%}'
PROMPT='
${user_and_pc}${prompt_git_home}%{$reset_color%}${my_prompt}$(git_prompt_info)%{$reset_color%}'

RPROMPT='%{$c1%}-%?- %*%{$reset_color%}'

#  vim: set ft=zsh sw=2 et tw=0:
```

**Comment by mcornella** [comment]
Ok, I can't see anything obvious. Please post the output of the following commands:
```sh
zle -l
```
and
```sh
which -a echo
```

**Comment by jbulchergist** [comment]
```
$ which -a echo                                                                                                 -0- 13:17:14
echo () {
        builtin zle .beep && _zsh_highlight
}
echo: shell built-in command
/usr/bin/echo
/bin/echo
/usr/bin/X11/echo
/usr/bin/echo
/bin/echo
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
```

```
$ zle -l
_bash_complete-word -C .complete-word _bash_completions
_bash_list-choices -C .list-choices _bash_completions
_complete_debug -C .complete-word _complete_debug
_complete_help -C .complete-word _complete_help
_complete_tag -C .complete-word _complete_tag
_correct_filename -C .complete-word _correct_filename
_correct_word -C .complete-word _correct_word
_expand_alias -C .complete-word _expand_alias
_expand_word -C .complete-word _expand_word
_history-complete-newer -C .complete-word _history_complete_word
_history-complete-older -C .complete-word _history_complete_word
_list_expansions -C .list-choices _expand_word
_most_recent_file -C .complete-word _most_recent_file
_next_tags -C .list-choices _next_tags
_read_comp -C .complete-word _read_comp
accept-and-hold
accept-and-infer-next-history
accept-and-menu-complete
accept-line
accept-line-and-down-history
accept-search
argument-base
auto-suffix-remove
auto-suffix-retain
backward-delete-char
backward-delete-word
backward-kill-line
backward-kill-word
backward-word
beep
beginning-of-buffer-or-history
beginning-of-history
beginning-of-line
beginning-of-line-hist
capitalize-word
clear-screen
complete-word
copy-prev-shell-word
copy-prev-word
copy-region-as-kill
delete-char
delete-char-or-list
delete-word
describe-key-briefly
digit-argument
down-case-word
down-history
down-line-or-search
edit-command-line
emacs-backward-word
emacs-forward-word
end-of-buffer-or-history
end-of-history
end-of-line
end-of-line-hist
end-of-list
exchange-point-and-mark
execute-last-named-cmd
execute-named-cmd
expand-cmd-path
expand-history
expand-or-complete
expand-or-complete-prefix
expand-or-complete-with-dots
expand-word
forward-word
get-line
gosmacs-transpose-chars
history-beginning-search-backward
history-beginning-search-backward-end (history-search-end)
history-beginning-search-forward
history-beginning-search-forward-end (history-search-end)
history-incremental-pattern-search-backward
history-incremental-pattern-search-forward
history-incremental-search-backward
history-incremental-search-forward
history-search-backward
history-search-forward
history-substring-search-down
history-substring-search-up
infer-next-history
insert-last-word
kill-buffer
kill-line
kill-region
kill-whole-line
kill-word
list-choices
list-expand
magic-space
menu-complete
menu-expand-or-complete
menu-select
neg-argument
orig-complete-word -C .complete-word _main_complete
orig-delete-char-or-list -C .delete-char-or-list _main_complete
orig-expand-or-complete -C .expand-or-complete _main_complete
orig-expand-or-complete-prefix -C .expand-or-complete-prefix _main_complete
orig-list-choices -C .list-choices _main_complete
orig-menu-complete -C .menu-complete _main_complete
orig-menu-expand-or-complete -C .menu-expand-or-complete _main_complete
orig-menu-select -C .menu-select _main_complete
orig-reverse-menu-complete -C .reverse-menu-complete _main_complete
overwrite-mode
pound-insert
prepend-sudo
push-input
push-line
push-line-or-edit
quote-line
quote-region
quoted-insert
read-command
recursive-edit
redisplay
redo
reset-prompt
reverse-menu-complete
self-insert (ordinary-key-press)
self-insert-unmeta
send-break
set-local-history
set-mark-command
spell-word
split-undo
transpose-chars
transpose-words
undefined-key
undo
universal-argument
up-case-word
up-history
up-line-or-search
vi-add-eol
vi-add-next
vi-backward-blank-word
vi-backward-char
vi-backward-delete-char
vi-backward-kill-word
vi-backward-word
vi-beginning-of-line
vi-caps-lock-panic
vi-change
vi-change-eol
vi-change-whole-line
vi-cmd-mode
vi-delete
vi-delete-char
vi-digit-or-beginning-of-line
vi-down-line-or-history
vi-end-of-line
vi-fetch-history
vi-find-next-char
vi-find-next-char-skip
vi-find-prev-char
vi-find-prev-char-skip
vi-first-non-blank
vi-forward-blank-word
vi-forward-blank-word-end
vi-forward-char
vi-forward-word
vi-forward-word-end
vi-goto-column
vi-goto-mark
vi-goto-mark-line
vi-history-search-backward
vi-history-search-forward
vi-indent
vi-insert
vi-insert-bol
vi-join
vi-kill-eol
vi-kill-line
vi-match-bracket
vi-open-line-above
vi-open-line-below
vi-oper-swap-case
vi-pound-insert
vi-put-after
vi-put-before
vi-quoted-insert
vi-repeat-change
vi-repeat-find
vi-repeat-search
vi-replace
vi-replace-chars
vi-rev-repeat-find
vi-rev-repeat-search
vi-set-buffer
vi-set-mark
vi-substitute
vi-swap-case
vi-undo-change
vi-unindent
vi-up-line-or-history
vi-yank
vi-yank-eol
vi-yank-whole-line
what-cursor-position
where-is
yank
yank-pop
zle-line-finish
zle-line-init
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
echo:zle: widgets can only be called when ZLE is active
```

**Comment by mcornella** [comment]
```sh
echo () {
        builtin zle .beep && _zsh_highlight
}
```
There you go, echo is a function which uses zle, and that is not allowed unless the function is a zle widget. Where is this defined? You can quickly test that this is the culprit by deleting the function with `unfunction echo`

**Comment by jbulchergist** [comment]
The theme works after I `unfunction echo`. I have no idea where `echo` is defined. I certainly didn't do it, and it isn't defined in `.oh-my-zsh`. I wonder if its an openSUSE thing (I'm still learning the system)? Any idea on how to fix it, besides running `unfunction echo` in `.zshrc`?

**Comment by apjanke** [comment]
Let's look at your rc files and figure out where the definition is coming from. Grab [this diagnostic script](https://github.com/apjanke/oh-my-zsh-custom/blob/master/diagnostics.zsh), source it, run `omz_diagnostic_dump`, and post the output as a gist. That'll grab copies of all your relevant zsh config files. We may be able to help you track it down from there.

**Comment by jbulchergist** [comment]
Here you go: https://gist.github.com/jbulchergist/c3592693d18b5b6dac63

Note: when I ran the diagnostic dump, the command complained, `_omz_diagnostic_dump_one_big_text:93: command not found: infocmp`

Also, I discovered that I needed to run `unfunction echo` after `source $ZSH/oh-my-zsh.sh` (line 663 in the gist). If I ran it before, it would complain `unfunction: no such hash table element: echo`.

**Comment by apjanke** [comment]
Oh yeah... if basic things like `echo` are redefined, that's going to affect the debugging script. I'll have to do something about that.

Thanks! I'll have a look through it when I have some time.

**Comment by apjanke** [comment]
Okay, could you also post these files too, since they're sourced from the /etc/z* or `~/.zshrc` rc files?
* `/etc/profile`
* `/etc/bash.bashrc`
* `~/.zsh_aliases`
* `~/.zsh_personal`
I think you can just add them as additional files to your existing gist.

**Comment by mcornella** [comment]
> Okay, could you also post these files too, since they're sourced from the /etc/z* or ~/.zshrc rc files?
> 
>  - `/etc/profile`
>  - `/etc/bash.bashrc`
>  - `~/.zsh_aliases`
>  - `~/.zsh_personal`

@jbulchergist:
> Also, I discovered that I needed to run unfunction echo after source $ZSH/oh-my-zsh.sh (line 663 in the gist). If I ran it before, it would complain unfunction: no such hash table element: echo.

This is guarantee that `/etc/*` files are not to blame. But me too I'm suspectful of those `~/.zsh_*` files. The diagnostic dump you posted also confirms that there are no custom files defining this. So, to reiterate, please post these files:

- `~/.zsh_aliases`
- `~/.zsh_personal`

-----
@apjanke: you can use `builtin echo` to force using the "good" echo, and so on for other _builtins_. For commands, you can use `command`.

**Comment by apjanke** [comment]
I'd still like to see all the files, just to get a complete picture of his configuration. (And for my benefit too, to help decide whether to add them or some `source`-chasing logic to the diagnostic dump function.)

**Comment by mcornella** [comment]
Good point, can't argue with that :smile:

**Comment by apjanke** [comment]
Okay, I've updated that `omz_diagnostic_dump` function with some defensive programming to handle redefinition of `echo` and other core commands. It should now run correctly even if they're defined weirdly, and also detect their redefinition and include that in the diagnostic dump file.

@jbulchergist, would you mind running the new version without doing `unfunction echo` first and see if it a) works for you and b) catches the fact that `echo` was redefined? (Worked for me in my test but I'd like to see confirmation in the wild.)

**Comment by jbulchergist** [comment]
Sorry for the blackout. I don't have access to the `.zsh_*` files, but I wrote them myself and they don't define an `echo` function. I'll try to post them when I get a chance, but the `unfunction` trick works on the line after sourcing `oh-my-zsh.sh`. The only reason I placed it after the other two files are sources was so I could keep the file organized better. In other words, sourcing the `oh-my-zsh.sh` script is creating the `echo` function.

```sh
source $ZSH/oh-my-zsh.sh
source ~/.zsh_aliases
source ~/.zsh_personal
 
# The echo functio was abducted from the shell, causing issues with my theme.
# Return control back to the shell built-in.
unfunction echo
```

I'll also try running the new version as soon as I get a chance.

**Comment by apjanke** [comment]
Okay, well. Now we've got a mystery I'm excited about. This is very curious.

There's nothing in OMZ itself that redefines echo, at least as far as I can see from grepping the whole source tree for "echo". Unless maybe something is doing a dynamic eval or something.

I think the next place to look is in the plugins you're loading, to see if they're pulling in anything external that redefines echo. I'll have a look when I'm back at a computer.

As a test of this hypothesis, could you turn off all your plugins and see if "echo" still gets overridden? (Just put a "which echo" after the "source oh-my-zsh.sh" in your ~/.zshrc.) And if that does it, then turn them back on one by one, to pinpoint the culprit.

Andrew

**Comment by mcornella** [comment]
Ok, at this point I'd use the big guns. Please run the command in this [Troubleshooting page](https://github.com/robbyrussell/oh-my-zsh/wiki/Troubleshooting#everything-else) and post the file in a gist.

**Comment by jbulchergist** [comment]
I found the culprit - it was the plugin `history-substring-search`. When I take it out of the plugins array, I find that I no longer needed `unfuction echo`.

I've gotten accustomed to *ctrl+r* for history searches since I bounce back and forth between bash, so loosing this plugin isn't a big deal for me personally. Still, if you want me to help fix this issue, I can run code for you and tell you the results.

**Comment by apjanke** [comment]
I would still be curious to see that debugging output, if you have time. I'm not able to reproduce the issue (on OS X or Debian), even running the exact same set of plugins as you are. This is a really interesting problem, and figuring out the "why" could help with other stuff. (Or at least my own understanding.)

Based on what your `echo` got redefined as, I'll bet it's [line 255 in `history-substring-search.zsh`](https://github.com/robbyrussell/oh-my-zsh/blob/aa16e18644b3fee0b3cf0320692d6566d539cdd0/plugins/history-substring-search/history-substring-search.zsh#L255) that's doing the redefinition. But I don't know what it would be interacting with such that it thought `echo` was one of the widgets it should redefine.

**Comment by jbulchergist** [comment]
OK, here's the omz-debug.log from `zsh -xv &> >(tee ~/omz-debug.log 2>/dev/null)`

https://gist.github.com/jbulchergist/a7d64657bdadfdd6c6ae

**Comment by apjanke** [comment]
Thank you, John! I'm stumped, but this is going to be interesting to pore over for a while.

**Comment by mcornella** [comment]
Ok, I have a hunch but I still don't know the full details. Basically, this bug is a combination of both the `history-substring-search` plugin, which implements a very outdated version of [zsh-syntax-highlighting](https://github.com/zsh-users/zsh-syntax-highlighting) (culprit being [line 273](https://github.com/robbyrussell/oh-my-zsh/blob/aa16e18644b3fee0b3cf0320692d6566d539cdd0/plugins/history-substring-search/history-substring-search.zsh#L273) of the plugin); and a file in `/etc/profile.d` directory which defines a whole bunch of aliases, one of them being `alias 'beep=echo -en "\007"'` that I think makes the bug appear. Look at the file `/etc/profile.d/alias.bash` on lines 49 to 53. 

In a gist, the zsh-syntax-highlighting implementation in `history-substring-search` plugin defines, on line 273, a beep function:
```zsh
eval 'beep() { builtin zle .beep && _zsh_highlight } ; zle -N beep'
```
You can see this in your gist as well.

Then, since we had a beep alias defined on that `alias.bash` file – _and this is the part where I don't really know how it happens (maybe some obscure zsh setting or bug?)_ – a collision happens in which the `beep` function is turned into an `echo` function.

You can test my hypothesis in the following ways (either one of them, not both at once):

1. Go to lines 49 to 53 of `/etc/profile.d/alias.bash` and comment them (I presume you'll need to be root). If the bug is not triggered with the history-substring-search plugin enabled, I'm on the right track.

2. Add a _canary_; put in place a command to nail down the offending lines in the history-substring-search plugin. Specifically, write ```zsh which -a echo ``` on `plugins/history-substring-search/history-substring-search.zsh` before [line 253](https://github.com/robbyrussell/oh-my-zsh/blob/aa16e18644b3fee0b3cf0320692d6566d539cdd0/plugins/history-substring-search/history-substring-search.zsh#L253) and after [line 281](https://github.com/robbyrussell/oh-my-zsh/blob/aa16e18644b3fee0b3cf0320692d6566d539cdd0/plugins/history-substring-search/history-substring-search.zsh#L281) (before and after that big `for`).

You go try that and let me know if my hypothesis is correct.
**ProTip:** you can quickly reload the terminal by running `exec zsh`.

Cheers

**Comment by apjanke** [comment]
Wow. Woww. Nice sleuthing!

That looks like it. I've got it reproducing on OS X 10.9 and Debian 7 using your diagnosis now. Simply adding `alias 'beep=echo -en "\007"'` to `~/.zshrc` before sourcing OMZ with just the `history-substring-search` plugin is sufficient to get the breakage.

```
Last login: Thu May 28 19:00:48 on ttys001
echo:zle: widgets can only be called when ZLE is active
[in ~]
$ which echo
echo () {
	builtin zle .beep && _zsh_highlight
}
echo:zle: widgets can only be called when ZLE is active
[in ~]
$
```

I think the mechanism is pretty straightfoward, if unintuitive, if you make an assumption about zsh parsing. Consider:

    eval "$clean_event() { builtin zle $event && _zsh_highlight } ; zle -N $clean_event"

Inside that string that's being eval'ed, `$clean_event` gets substituted with `beep`. Then it's evaled as its own statement. If `(` is a token delimiter, then `beep` is in "command position" and is eligible for alias expansion, which is a simple string substitution. That would give you this.

    echo -en "\007"() { builtin zle $event && _zsh_highlight } ; zle -N $clean_event }

Which looks nonsensical, but whaddya know, it's a valid syntax for defining a function, or at least `zsh` thinks it is. (!!!) Example with bare `zsh` (no `~/.zshrc`):

````
Last login: Thu May 28 19:13:17 on ttys003
eilonwy% which echo
echo: shell built-in command
eilonwy% echo -en "\007"() { printf foobar }
eilonwy% which echo
echo () {
	printf foobar
}
eilonwy% echo
foobar%                                                                         eilonwy%
```

I... did not see that coming.

This may be part of why the core `zsh` documentation recommends using functions instead of aliases in all cases.

I think maybe it can be fixed by just sticking an explicit `function` in the function definition, so `$clean_event` is no longer in command position and is not eligible for alias expansion.

````
eval "function $clean_event () { builtin zle $event && _zsh_highlight } ; zle -N $clean_event"
````

And if that's the case, maybe we should be using explicit `function` everywhere in OMZ as part of our coding style, as defense against this alias-collision bug.

**Comment by apjanke** [comment]
Okay, check out [Complex commands in the ZSH ref manual](http://zsh.sourceforge.net/Doc/Release/Shell-Grammar.html#Complex-Commands) and scroll down to "function". It lets you specify any number of words to be alternate names to a function. So the presence of `()` in a line basically takes everything leading up to it and redefines it as a function. (??)

````
Last login: Thu May 28 19:28:11 on ttys002
eilonwy% echo Here -is -a 'bunch of junk'() { printf DARNIT }
eilonwy% echo Hello
DARNIT%                                                                         eilonwy% which echo
echo () {
	printf DARNIT
}
eilonwy% which -- -is
-is () {
	printf DARNIT
}
eilonwy% which -- 'bunch of junk'
'bunch of junk' () {
	printf DARNIT
}
eilonwy%
````

I guess you have to be really careful with lines that have `()` and `{...}` in them. And aliasing opens up a whole can of worms. I think this means it would have broken if `echo` or something else important had been *any* of the words in the `beep` alias, not just the first.

**Comment by apjanke** [comment]
Fun ZSH tricks: this one-liner does not actually print out some math tidbits.

    echo for k [ 1 .. 5 ] map builtin . fact() printf "$k"

**Comment by mcornella** [comment]
Ok, so that's what's happening... To sum up, [here's the relevant section from zsh documentation](http://zsh.sourceforge.net/Doc/Release/Shell-Grammar.html#Aliasing):

> Note also the unhelpful interaction of aliases and function definitions:
>
> ```zsh
> alias func='noglob func'
> func() {
>     echo Do something with $*
> }
> ```
>
> Because aliases are expanded in function definitions, this causes the following command to be executed:
> 
> ```zsh
> noglob func() {
>     echo Do something with $*
> }
> ```
>
> which defines noglob as well as func as functions with the body given. **To avoid this, either quote the name func or use the alternative function definition form ‘function func’.**

So we'll have to use the `function func` form:
```zsh
function beep { ... }
```
or quote the function name:
```zsh
'beep' () { ... }
```

Fun fact: if the alias is global (`alias -g beep=echo`), the `function func` form will still be wrong.
Another option we have, but with potential side effects, is [disabling alias expansion](http://zsh.sourceforge.net/Doc/Release/Options.html#index-ALIASES) and enabling it again after loading omz, with `setopt no_aliases`.

**Comment by mcornella** [comment]
@jbulchergist we found the culprit, please update the title to `history-substring-search defines function that collides with beep alias` (or something similar), and try the following fix:

```diff
diff --git a/plugins/history-substring-search/history-substring-search.zsh b/plugins/history-substring-search/history-substring-search.zsh
index 22f03dd..45c0280 100644
--- a/plugins/history-substring-search/history-substring-search.zsh
+++ b/plugins/history-substring-search/history-substring-search.zsh
@@ -252,11 +252,11 @@ if [[ $+functions[_zsh_highlight] -eq 0 ]]; then
   # Override ZLE widgets to make them invoke _zsh_highlight.
   for event in ${${(f)"$(zle -la)"}:#(_*|orig-*|.run-help|.which-command)}; do
     if [[ "$widgets[$event]" == completion:* ]]; then
-      eval "zle -C orig-$event ${${${widgets[$event]}#*:}/:/ } ; $event() { builtin zle orig-$event && _zsh_highlight } ; zle -N $event"
+      eval "zle -C orig-$event ${${${widgets[$event]}#*:}/:/ } ; function $event { builtin zle orig-$event && _zsh_highlight } ; zle -N $event"
     else
       case $event in
         accept-and-menu-complete)
-          eval "$event() { builtin zle .$event && _zsh_highlight } ; zle -N $event"
+          eval "function $event { builtin zle .$event && _zsh_highlight } ; zle -N $event"
           ;;
 
         # The following widgets should NOT remove any previously
@@ -270,7 +270,7 @@ if [[ $+functions[_zsh_highlight] -eq 0 ]]; then
             (completion|user):*)
               ;;
             *)
-              eval "$clean_event() { builtin zle $event && _zsh_highlight } ; zle -N $clean_event"
+              eval "function $clean_event { builtin zle $event && _zsh_highlight } ; zle -N $clean_event"
               ;;
           esac
           ;;
```

To apply the patch, navigate to the oh-my-zsh folder, run `git apply` and paste the patch above, then press `CTRL`+`D` to apply the patch.

**Comment by apjanke** [comment]
I tested your patch on OS X 10.9 and it looks good.

I also updated the [Coding style guide wiki page](https://github.com/robbyrussell/oh-my-zsh/wiki/Coding-style-guide) to call for explicit use of `function` and link to that Aliases doco to explain why.

**Comment by jbulchergist** [comment]
The patch works for me as well. Should I leave the issue open till the patch is applied to the repo?

**Comment by mcornella** [comment]
Yes, leave it open. Let me work on a PR and I'll auto-close this one when it is merged.
Thanks!

**Comment by apjanke** [comment]
It looks like upstream for this plugin is actually https://github.com/zsh-users/zsh-history-substring-search, and they've pulled in a more recent version of the zsh-syntax-highlighting code which fixes this. Maybe we should just pull the current version of zsh-history-substring-search in to the OMZ plugin. 

Thoughts? @sunaku, @sorin-ionescu – is the HEAD version of zsh-history-substring-search good to use?

**Comment by sorin-ionescu** [comment]
HEAD is fine.

**Comment by sunaku** [comment]
:+1: Agreed.  The master branch of zsh-history-substring-search has been stable for a long time. :older_man:

**Comment by apjanke** [comment]
Great.

Hey, your new work on zsh-history-substring-search is good stuff, especially the documentation on key bindings.

I wrote [a little script to automatically do the update from upstream](https://github.com/apjanke/oh-my-zsh/blob/1ae2dcb5135cd3d09e55a9602e80862c8331bf5f/plugins/history-substring-search/update-from-upstream.zsh) and add OMZ-specific stuff to it, and used it to update OMZ to the current HEAD. It's on my branch https://github.com/apjanke/oh-my-zsh/tree/histsubstr-update-2015_04_26; no PR for it yet.

The OMZ-specific stuff is:
* An additional section to the README that explains the relationship between the OMZ-distributed plugin and your upstream repo, and includes a direct link to your repo.
* We bind the up and down arrow keys in the plugin by default, instead of requiring the user to do it in `~/.zshrc`. This is a matter of convenience, and matches the existing behavior for the OMZ plugin.

If `history-substring-search` users are okay with it, I think we should do the keymap-specific bindings (as documented in the upstream README) by default as well in the OMZ plugin layer.

**Comment by apjanke** [comment]
I put in a PR to do the updating from upstream: #4190.

**Comment by mcornella** [comment]
Fixed by #4190
